### PR TITLE
fix(route53): vpc region in template overridden by stack region

### LINKED ISSUE #20496: aws_Route53: (HostedZone.add_vpc set incorrect region)
### Describe the bug

route53.HostZone.add_vpc sets regions for the stack's regions, not the VPC's region.

### Expected Behavior

The regions for the VPCs should be correct.

```
  "PrivateDnsE0FF4F9B": {
   "Type": "AWS::Route53::HostedZone",
   "Properties": {
    "Name": "slurmiad.local.",
    "VPCs": [
     {
      "VPCId": "vpc-nnnnnnnnnnnnnnnnn",
      "VPCRegion": "us-east-1"
     },
     {
      "VPCId": "vpc-nnnnnnnnnnnnnnnnn",
      "VPCRegion": "us-west-2"
     },
     {
      "VPCId": "vpc-nnnnnnnnnnnnnnnnn",
      "VPCRegion": "eu-west-1"
     }
    ]
   },
   "Metadata": {
    "aws:cdk:path": "slurmiad/PrivateDns/Resource"
   }
  },
```

### Current Behavior

The resulting CFN template has the correct VpcIds, but the incorrect regions:

```
  "PrivateDnsE0FF4F9B": {
   "Type": "AWS::Route53::HostedZone",
   "Properties": {
    "Name": "slurmiad.local.",
    "VPCs": [
     {
      "VPCId": "vpc-nnnnnnnnnnnnnnnnn",
      "VPCRegion": "us-east-1"
     },
     {
      "VPCId": "vpc-nnnnnnnnnnnnnnnnn",
      "VPCRegion": "us-east-1"
     },
     {
      "VPCId": "vpc-nnnnnnnnnnnnnnnnn",
      "VPCRegion": "us-east-1"
     }
    ]
   },
   "Metadata": {
    "aws:cdk:path": "slurmiad/PrivateDns/Resource"
   }
  },
```

### Reproduction Steps

I'm creating a Route53.HostedZone for use in 3 VPCs that are located in 3 different regions.
The VPCs aren't part of the stack and are created in CDK using 

```
                ec2.Vpc.from_lookup(
                    self, f"Vpc{region_dict['Region']}",
                    region = region_dict['Region'],
                    vpc_id = region_dict['VpcId'])
```

This causes an update to cdk.context.json where the VPC ids and regions are correct.
Extract from cdk.context.json:

```
  "vpc-provider:account=nnnnnnnnnnnn:filter.vpc-id=vpc-nnnnnnnnnnnnnnnnn:region=us-west-2:returnAsymmetricSubnets=true": {
    "vpcId": "vpc-nnnnnnnnnnnnnnnnn",
    "vpcCidrBlock": "10.3.0.0/16",
    "availabilityZones": [],
    "subnetGroups": [
      {
        "name": "Private",
        "type": "Private",
        "subnets": [
          {
            "subnetId": "subnet-nnnnnnnnnnnnnnnnn",
            "cidr": "10.3.64.0/18",
            "availabilityZone": "us-west-2a",
            "routeTableId": "rtb-nnnnnnnnnnnnnnnnn"
          },
          {
            "subnetId": "subnet-nnnnnnnnnnnnnnnnn",
            "cidr": "10.3.128.0/18",
            "availabilityZone": "us-west-2b",
            "routeTableId": "rtb-nnnnnnnnnnnnnnnnn"
          },
          {
            "subnetId": "subnet-nnnnnnnnnnnnnnnnn",
            "cidr": "10.3.192.0/18",
            "availabilityZone": "us-west-2c",
            "routeTableId": "rtb-nnnnnnnnnnnnnnnnn"
          }
        ]
      },
      {
        "name": "Public",
        "type": "Public",
        "subnets": [
          {
            "subnetId": "subnet-nnnnnnnnnnnnnnnnn",
            "cidr": "10.3.0.0/26",
            "availabilityZone": "us-west-2a",
            "routeTableId": "rtb-nnnnnnnnnnnnnnnnn"
          },
          {
            "subnetId": "subnet-nnnnnnnnnnnnnnnnn",
            "cidr": "10.3.0.64/26",
            "availabilityZone": "us-west-2b",
            "routeTableId": "rtb-nnnnnnnnnnnnnnnnn"
          },
          {
            "subnetId": "subnet-nnnnnnnnnnnnnnnnn",
            "cidr": "10.3.0.128/26",
            "availabilityZone": "us-west-2c",
            "routeTableId": "rtb-nnnnnnnnnnnnnnnnn"
          }
        ]
      }
    ]
  },
```

I create the Hosted Zone:

```
            self.hosted_zone = route53.HostedZone(self, "PrivateDns",
                vpcs = [self.vpc],
                zone_name = self.config['Domain']
            )
            self.hosted_zone.add_vpc(remote_vpcs[region_dict['Region']])
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### CDK CLI Version

2.25.0

### Framework Version

_No response_

### Node.js Version

16.15.0

### OS

AmazonLinux2

### Language

Python

### Language Version

Python 3.7.10

### Other information

_No response_

### PR DESCRIPTION
**Pull request by daschaa**
Fixes #20496 

This PR implements the proposed change in #20496 - When a region is set in the vpc it is used in the CloudFormation template. Otherwise the region from the respective stack is used.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by daschaa** [comment]
@peterwoodworth Since you commented on the original issue - could you review this PR please? I would appreciate the feedback! ❤️

**Comment by peterwoodworth** [comment]
Sure thing @daschaa, I'll try to get to this today 🙂

**Review by peterwoodworth — DISMISSED** [review]
Thanks for the contribution @daschaa!

This is going to be a slightly larger PR than I had anticipated. We need to add some additional functionality elsewhere for this to work.

First, let's take a look at how our S3 module handles this:
[`Bucket.fromBucketAttributes()`](https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-s3.Bucket.html#static-fromwbrbucketwbrattributesscope-id-attrs) has a [`region`](https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-s3.BucketAttributes.html#region) prop the user can pass in for when the region of the stack and bucket differ. This ends up being represented to the user through `IBucket.env.region`
https://github.com/aws/aws-cdk/blob/ab288785279bf3d70ad3329e27ad1f6d2397942e/packages/%40aws-cdk/aws-s3/lib/bucket.ts#L1588-L1591

`Vpc.fromVpcAttributes()` and `Vpc.fromLookup()` should both support this. However, neither of these methods currently support this. The former doesn't allow you to set the region at all, and the latter does, but it isn't set [when the construct is created](https://github.com/aws/aws-cdk/blob/ab288785279bf3d70ad3329e27ad1f6d2397942e/packages/%40aws-cdk/aws-ec2/lib/vpc.ts#L1176). Both of these functionalities will need to be implemented 🙂 

As for tests, let's create tests in the EC2 library to verify that the region is properly set with these methods. Additionally, for the hostedzone test, could we use one of these import methods instead of a mock construct?

**Comment by peterwoodworth** [comment]
No worries @daschaa 

> it doesn't really test the functionality - just if it is passed correctly

Well, that is all we need to test here! Right? All we want to know is that the Vpc you create with `fromLookup` sets its `.env.region` prop to the one you pass in. Check out [this test file](https://github.com/aws/aws-cdk/blob/main/packages/@aws-cdk/aws-ec2/test/vpc.from-lookup.test.ts) - this is the file where you'll want to add the test and it has plenty of examples 🙂 

Let me know if you need any more help!

**Comment by daschaa** [comment]
@peterwoodworth I can now continue working on this feature again. Sorry for the delay!
I have added the region prop the the `fromLookup` and `fromVpcAttributes` and used `fromVpcAttributes` in the test for the hosted zone. 
However I'm not really happy about the typing of the `LookedUpVpc`. I extended the `props`-property with region and decided to not introduce an own type. I thought it would be overkill to just implement it for the region prop. Is it ok like that? I would be happy to get feedback from your side and maybe some better ideas about the typing. Thanks in advance!

**Comment by daschaa** [comment]
@peterwoodworth I just saw that I did not rebase this branch very well, but I think everything is fine. At least the changes still look good :) So please excuse me for this huge list of commits in this PR 😄

**Comment by TheRealAmazonKendra** [comment]
Rebasing gets weird sometimes ¯\\\_(ツ)\_/¯

**Comment by daschaa** [comment]
@mergifyio update

**Comment by daschaa** [comment]
@TheRealAmazonKendra Is this mergify command new? Never saw this before and wondered if we should use it instead of the GitHub Update Branch command :)

**Comment by TheRealAmazonKendra** [comment]
> @TheRealAmazonKendra Is this mergify command new? Never saw this before and wondered if we should use it instead of the GitHub Update Branch command :)

I use it because it doesn't dismiss previous reviews and will rerun workflows. I think I'm the only one who really uses the Mergify commands, though. Either way is fine.

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 1]
```diff
@@ -1983,7 +1995,9 @@ class ImportedVpc extends VpcBase {
   private readonly cidr?: string | undefined;
 
   constructor(scope: Construct, id: string, props: VpcAttributes, isIncomplete: boolean) {
-    super(scope, id);
+    super(scope, id, {
+      region: props.region ?? undefined,
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts:1999]: This is redundant. `region: props.region` means the same thing as `region: props.region ?? undefined` when `props.region` is potentially undefined.

```suggestion
      region: props.region,
```

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 2]
```diff
@@ -2032,8 +2046,10 @@ class LookedUpVpc extends VpcBase {
   public readonly isolatedSubnets: ISubnet[];
   private readonly cidr?: string | undefined;
 
-  constructor(scope: Construct, id: string, props: cxapi.VpcContextResponse, isIncomplete: boolean) {
-    super(scope, id);
+  constructor(scope: Construct, id: string, props: cxapi.VpcContextResponse & { region?: string }, isIncomplete: boolean) {
+    super(scope, id, {
+      region: props.region ?? undefined,
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts:2051]: Same comment as above. 
```suggestion
      region: props.region,
```

**Review by TheRealAmazonKendra — DISMISSED** [review]
I've added some very preliminary feedback here. I'd also like to see an integration test for this change, please.

Oh, also, please take a look and see if you think there's a place where README documentation will need to be updated with this change. I'll trust your judgement on that.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Actually, I'm going to revise my earlier statement about using `@Mergifyio update`. Please do use it instead of merging via other methods. I'm finding it much easier to track when changes have been made on the PRs where that command is being used. Since we're trying to provide faster responses to PRs in general, this helps us be more efficient. (Also just putting this back into changes requested so that I'll see when you've made a code update).

**Comment by daschaa** [comment]
@TheRealAmazonKendra I have added your suggested changes. Thanks for the feedback!
Regarding the integration tests: Do you mean Integration Tests for the `fromLookup` and `fromVpcAttributes` methods? I wonder if this would be easily possible since the snapshot would not do a real lookup, right? 🤔 
Or do you mean for the `addVpc` method?

**Comment by TheRealAmazonKendra** [comment]
New/updated integration tests actually create the resources and deploy them. They just don't run in a public account so you never see it happen. We also will see from the synthesized template that the region assigned to the VPC will be different from the one you assign for the stack. If you create the test using the `IntegTest` construct and then run `yarn integ --update-on-failed <filename>.js` then it will setup the resources, tear them back down, and produce the expected snapshot files to compare against later. Check out our new documentation on [integration tests](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md) for more info.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
I thought I commented this earlier but I can't find it anywhere. I tend to think that region should be added to `VpcContextResponse` and added to `attributes` more organically that way, instead of through the additional prop.

**Comment by TheRealAmazonKendra** [comment]
> @TheRealAmazonKendra I have added your suggested changes. Thanks for the feedback! Regarding the integration tests: Do you mean Integration Tests for the `fromLookup` and `fromVpcAttributes` methods? I wonder if this would be easily possible since the snapshot would not do a real lookup, right? 🤔 Or do you mean for the `addVpc` method?

I just realized that I only kind of answered this question. I'd want to see integ tests on `fromLookup` and `fromVpcAttributes`. How I see that working is to create two stacks in your test. One in region-a that creates the VPC and one in region-b that uses `fromLookup` and `fromVpcAttributes` to import it. Only list the stack that uses these functions in your `testCases` array.

**Comment by daschaa** [comment]
@TheRealAmazonKendra Sorry for not working on this consistently - I have a lot on my plate right now.
I have now added the integration tests for `fromLookup` and `fromVpcAttributes` as you suggested and I moved the region prop to the `VpcContextResponse`. 
I look forward to your feedback :)

**Comment by TheRealAmazonKendra** [comment]
> @TheRealAmazonKendra Sorry for not working on this consistently - I have a lot on my plate right now. I have now added the integration tests for `fromLookup` and `fromVpcAttributes` as you suggested and I moved the region prop to the `VpcContextResponse`. I look forward to your feedback :)

No need to be sorry at all, you're contributing your time and resources to this and we're super grateful for it!

**Comment by TheRealAmazonKendra** [comment]
FYI - I haven't forgot this PR but I do think there's more refactoring that needs to be done so I'm taking a deeper dive into the code to make sure my suggestions aren't off base.

**Comment by daschaa** [comment]
@TheRealAmazonKendra Any updates on this one? :)

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]

**Comment by daschaa** [comment]
@Mergifyio update

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 3]
```diff
@@ -1203,7 +1210,12 @@ export class Vpc extends VpcBase {
       dummyValue: undefined,
     }).value;
 
-    return new LookedUpVpc(scope, id, attributes || DUMMY_VPC_PROPS, attributes === undefined);
+    const lookedUpVpcProps = {
+      ...attributes,
+      region: options.region,
```
[the remark this answers is not in the document]
Comment (daschaa) [reply packages/@aws-cdk/aws-ec2/lib/vpc.ts]: I think you are right, this assignment is redundant. I have now removed it and the integration tests still work as expected. Thanks for pointing this out :)

**Review by TheRealAmazonKendra — APPROVED** [review]
Thanks for all your work on this and patience in the long delay between feedback. This looks great!

**Comment by michaelciccarelli** [comment]
Can anyone follow up on what I'm seeing? I'm running 2.44.0 and this doesn't seem like it's working yet:

       if env.region == "us-east-1" and not self.bcp_env:
            useast2vpcid = os.getenv("VPC_ID_US_EAST_2", "MISSING__VPC_ID_US_EAST_2_FOR_HIGHAVAILABILITY")
            useast2vpc = Vpc.from_lookup(self, "vpcLookup-useast2", vpc_id=useast2vpcid, region="us-east-2")
            my_hosted_zone = route53.HostedZone(self, APP_NAME + "-route53",
                zone_name = subdomain + ".essent.us",
                vpcs = [ self.vpc,  useast2vpc  ]
            )

When I run it, the synth is fine but apply shows that it CDKformation doesn't have permission to add this vpc.. the issue is the vpc is showing up as in us-east-1 even though we're specifying us-east-2.. The following code fixes the issue with calling Cfn directly.... 

       if env.region == "us-east-1" and not self.bcp_env:
            useast2vpcid = os.getenv("VPC_ID_US_EAST_2", "MISSING__VPC_ID_US_EAST_2_FOR_HIGHAVAILABILITY")
#            useast2vpc = Vpc.from_lookup(self, "vpcLookup-useast2", vpc_id=useast2vpcid, region="us-east-2")
#            my_hosted_zone = route53.HostedZone(self, APP_NAME + "-route53",
            my_hosted_zone = route53.CfnHostedZone(self, APP_NAME + "-route53",
                name = subdomain + ".essent.us",
#                vpcs = [ self.vpc ]
                vpcs = [
                    route53.CfnHostedZone.VPCProperty(vpc_id=self.vpc_id, vpc_region = env.region),
                    route53.CfnHostedZone.VPCProperty(vpc_id=useast2vpcid, vpc_region = "us-east-2")
                ]
            )

**Pull request merged**